aries_vcx/src/core/profile/modular_libs_profile.rs

```diff
@@ -21,18 +26,125 @@ use super::profile::Profile;
 #[allow(dead_code)]
 #[derive(Debug)]
 pub struct ModularLibsProfile {
-    pub(crate) wallet: Arc<dyn BaseWallet>,
-    pub(crate) anoncreds: Arc<dyn BaseAnonCreds>,
-    pub(crate) anoncreds_ledger_read: Arc<dyn AnoncredsLedgerRead>,
-    pub(crate) anoncreds_ledger_write: Arc<dyn AnoncredsLedgerWrite>,
-    pub(crate) indy_ledger_read: Arc<dyn IndyLedgerRead>,
-    pub(crate) indy_ledger_write: Arc<dyn IndyLedgerWrite>,
+    wallet: Arc<dyn BaseWallet>,
+    anoncreds: Arc<dyn BaseAnonCreds>,
+
+    // ledger reads
+    anoncreds_ledger_read: Arc<dyn AnoncredsLedgerRead>,
+    indy_ledger_read: Arc<dyn IndyLedgerRead>,
+
+    // ledger writes
+    anoncreds_ledger_write: Arc<RwLock<dyn AnoncredsLedgerWrite + Send + Sync>>,
+    indy_ledger_write: Arc<RwLock<dyn IndyLedgerWrite + Send + Sync>>,
+    taa_configurator: Arc<RwLock<dyn TaaConfigurator + Send + Sync>>,
+}
+
+// The Proxy structs encapsulates dealing with RwLock while implementing IndyLedgerWrite trait,
+// so that rest of the codebase can call trait methods below, without being concerned about obtaining
+// read/write locks. This pattern also enabled us introduce mutability of ledger write structures
+// (setting up TAA in runtime) without having to sync up codebase across the board due to newly
+// introduced RwLock.
+#[derive(Debug)]
+pub struct IndyLedgerWriteProxy {
+    indy_ledger_write: Arc<RwLock<dyn IndyLedgerWrite + Send + Sync>>,
+}
+
+#[derive(Debug)]
+pub struct AnoncredsLedgerWriteProxy {
+    anoncreds_ledger_write: Arc<RwLock<dyn AnoncredsLedgerWrite + Send + Sync>>,
+}
+
+#[async_trait]
+impl IndyLedgerWrite for IndyLedgerWriteProxy {
+    async fn publish_nym(
+        &self,
+        submitter_did: &str,
+        target_did: &str,
+        verkey: Option<&str>,
+        data: Option<&str>,
+        role: Option<&str>,
+    ) -> VcxCoreResult<String> {
+        self.indy_ledger_write
+            .read()
```

Comment:
As we discussed, I am not a fan of the overhead associated with acquiring the read lock on every write operation. (Nor am I a fan of having the ability to randomly change TAA during runtime.) It would be good if we managed to find a solution which avoids this downside.